ssl/quic/quic_port.c

```diff
@@ -647,9 +659,132 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
+{
+    WPACKET wpkt = {0};
+    BUF_MEM *buf_mem = BUF_MEM_new();
+
+    if (buf_mem == NULL || (token->is_retry != 0 && token->is_retry != 1))
+        return 0;
+
+    if (!WPACKET_init(&wpkt, buf_mem)
+        || !WPACKET_memset(&wpkt, token->is_retry, 1)
+        || !WPACKET_memcpy(&wpkt, &token->timestamp,
+                           sizeof(token->timestamp))
+        || !WPACKET_memcpy(&wpkt, &token->odcid.id_len,
+                           sizeof(token->odcid.id_len))
+        || !WPACKET_memcpy(&wpkt, &token->odcid.id, token->odcid.id_len)
+        || !WPACKET_memcpy(&wpkt, &token->rscid.id_len,
+                           sizeof(token->rscid.id_len))
+        || !WPACKET_memcpy(&wpkt, &token->rscid.id, token->rscid.id_len)
+        || !WPACKET_memcpy(&wpkt, &token->remote_addr_len,
+                           sizeof(token->remote_addr_len))
+        || !WPACKET_memcpy(&wpkt, token->remote_addr, token->remote_addr_len)
+        || !WPACKET_get_total_written(&wpkt, buffer_len)
+        || !WPACKET_finish(&wpkt)
+        || (*buffer = OPENSSL_realloc(buf_mem->data, *buffer_len +
+                                      QUIC_RETRY_INTEGRITY_TAG_LEN)) == NULL) {
+        WPACKET_cleanup(&wpkt);
+        BUF_MEM_free(buf_mem);
+        return 0;
+    }
+
+    *buffer_len += QUIC_RETRY_INTEGRITY_TAG_LEN;
+    OPENSSL_free(buf_mem);
+    return 1;
+}
+
+/**
+ * @brief Parses contents of a buffer into a validation token.
+ *
+ * VALIDATION_TOKEN should already be initalized. Does some basic sanity checks.
+ *
+ * @param token   Validation token to fill data in.
+ * @param buf     Buffer of previously marshaled validation token.
+ * @param buf_len Length of |buf|.
+ */
+static int parse_validation_token(QUIC_VALIDATION_TOKEN *token,
+                                  const unsigned char *buf, size_t buf_len)
+{
+    PACKET pkt;
+
+    if (buf == NULL || token == NULL)
+        return 0;
+
+    token->remote_addr = NULL;
+
+    if (!PACKET_buf_init(&pkt, buf, buf_len)
+        || !PACKET_copy_bytes(&pkt, &token->is_retry, sizeof(token->is_retry))
+        || !(token->is_retry == 0 || token->is_retry == 1)
+        || !PACKET_copy_bytes(&pkt, (unsigned char *)&token->timestamp,
+                              sizeof(token->timestamp))
+        || !PACKET_copy_bytes(&pkt, &token->odcid.id_len,
+                              sizeof(token->odcid.id_len))
+        || !PACKET_copy_bytes(&pkt, (unsigned char *)&token->odcid.id,
+                              token->odcid.id_len)
+        || !PACKET_copy_bytes(&pkt, &token->rscid.id_len,
+                              sizeof(token->rscid.id_len))
+        || !PACKET_copy_bytes(&pkt, (unsigned char *)&token->rscid.id,
+                              token->rscid.id_len)
+        || !PACKET_copy_bytes(&pkt, (unsigned char *)&token->remote_addr_len,
+                              sizeof(token->remote_addr_len))
+        || (token->remote_addr = OPENSSL_malloc(token->remote_addr_len)) == NULL
```

Comment:
You need to check the lengths to not exceed the sizes. Also better would be to use subpackets here.